teps/0060-remote-resource-resolution.md

```diff
@@ -319,19 +324,289 @@ Why should this TEP _not_ be implemented?
+  own github secret to fetch their tasks/pipelines with).
+
+### Custom Tasks
+
+Abandon any plans to implement something specific in Pipelines. Just tell users to write custom tasks
+that fetch from their source of choice.
+
+1. User submits PipelineRun
+
+```yaml
+kind: PipelineRun
+metadata:
+  name: foo-run
+spec:
+  tasks:
+  - taskSpec:
+      spec:
+        apiVersion: gitref.tekton.dev/v1alpha1
+        kind: RemoteFetch
+        spec: # embedded custom task spec, tep-0061. These fields can be anything the Resolver wants to support.
+          repo: https://github.com/tektoncd/catalog.git
+          ref: main
+          path: /tasks/git-clone/0.4/git-clone.yaml
+      # params and workspaces for resolved git-clone task
+      params:
+      - name: url
+        value: my-repo/foo/bar.git
+      workspaces:
+      - name: output
+      # ...
+```
+
+2. PipelineRun reconciler creates a `Run` object for this `taskSpec`:
+
+```yaml
+kind: Run
+spec:
+  spec:
+    apiVersion: gitref.tekton.dev/v1alpha1
+    kind: RemoteFetch
+    spec:
+      repo: https://github.com/tektoncd/catalog.git
+      ref: main
+      path: /tasks/git-clone/0.4/git-clone.yaml
+  params:
+  - name: url
+    value: my-repo/foo/bar.git
+  workspaces:
+  - name: output
+  # ...
+```
+
+2. Custom Task reconciler is waiting for `gitref.tekton.dev/v1alpha1.RemoteFetch`. Fetches
+remote resource based on parameters from embedded spec.
+
+3. Custom Task reconciler creates `TaskRun` with fetched resource, passing through params,
+workspaces, etc...
+
+4. Custom Task reconciler mirrors status changes, results, etc from the `TaskRun` it created
+back to the `Run` that Pipeline reconciler is watching.
+
+Pros:
+- Reuses existing Custom Tasks pattern.
+- Doesn't require any changes in Pipelines or specification effort at all.
+- Lifetime of fetched resource can be tied to owner reference - the `TaskRun` that the custom task reconciler
+creates can add an owner reference to the `Run` for this I think?
+- Tekton _could_ provide one of these for git
+- Resource resolution is performed declaratively.
+- Resolution is executed concurrently while the Pipelines controller does other things.
+- Flexible since Custom Tasks can do "whatever they want".
+    - Example: if a Custom Task is implementing an -as-code flow it can fetch a Pipeline at a specific commit __AND__ re-write any referenced tasks in that Pipeline so that they too reference the specific commit as well.
+- Pipelines not responsible for anything Resolver-specific.
+- Development burden of writing a reconciler _could_ be alleviated with a custom task SDK?
+
+Cons:
+- No consistent specification of behaviour, potentially varies for every custom task implementing remote fetch.
+- No validation of fetched content by Pipelines.
+- No opportunity to apply defaults by Pipelines since the Custom Task controller is responsible for creating the \*Runs.
```

Comment:
the above 2 cons could be mitigated if the custom task was only responsible for fetching the task/pipeline, and not for actually creating the run

(i think this is very similar to the next option, but instead of having a new CRD we could use a run)

Reply:
Added note to capture this.